Modified for flatlist (from #629)

xcarpentier opened the pull request:
From https://github.com/FaridSafi/react-native-gifted-chat/pull/629

xcarpentier commented:
I think we need to have an expo publish on every PR to test easely...

cooperka commented:
Thanks for cleaning this up @xcarpentier :beers: But fyi I probably won't have time to review this week, you may want to find someone else to review.

brunocascio commented:
Great work @xcarpentier !

I am in vacation now, but I saw part of the code. Is there a way to remove internal state and use only props un order to make this lib completely stateless. What do you think?

brunolemos commented on src/MessageContainer.js:
```diff
@@ -95,72 +86,66 @@ export default class MessageContainer extends React.Component {
-        {...invertibleScrollViewProps}
-        ref={(component) => (this._invertibleScrollViewRef = component)}
-      />
+      <View style={{ transform: [{ scaleY: -1 }, { perspective: 1280 }] }}>
```
Why this?

brunolemos commented on src/MessageContainer.js:
```diff
@@ -95,72 +86,66 @@ export default class MessageContainer extends React.Component {
           automaticallyAdjustContentInsets={false}
           initialListSize={20}
           pageSize={20}
+          ref="flatListRef"
```
String ref was deprecated a year ago, you should use callback refs only.

```jsx
ref={flatListRef =>
  if (flatListRef) {
    this.flatListRef = flatListRef
  }
}
```

brunolemos commented on src/MessageContainer.js:
```diff
@@ -95,72 +86,66 @@ export default class MessageContainer extends React.Component {
+          renderItem={this.renderRow}
+          renderHeader={this.renderFooter}
+          renderFooter={this.renderLoadEarlier()}
+          style={{ transform: [{ scaleY: -1 }, { perspective: 1280 }] }}
```
Why this?

brunolemos commented on src/MessageContainer.js:
```diff
@@ -172,8 +157,9 @@ const styles = StyleSheet.create({
-    justifyContent: 'flex-end',
+  headerWrapper: {
+    flex: 1,
+    transform: [{ scaleY: -1 }, { perspective: 1280 }],
```
Why this?

brunolemos commented on src/__tests__/__snapshots__/MessageContainer.test.js.snap:
```diff
@@ -9,21 +9,70 @@ exports[`should render <MessageContainer /> and compare with snapshot 1`] = `
+    onScrollEndDrag={[Function]}
     pageSize={20}
-    renderFooter={[Function]}
+    renderFooter={null}
```
Why this?

brunolemos commented on src/__tests__/__snapshots__/MessageContainer.test.js.snap:
```diff
@@ -9,21 +9,70 @@ exports[`should render <MessageContainer /> and compare with snapshot 1`] = `
+            "scaleY": -1,
+          },
+          Object {
+            "perspective": 1280,
```
Why this?

brunolemos commented on src/__tests__/__snapshots__/MessageContainer.test.js.snap:
```diff
@@ -9,21 +9,70 @@ exports[`should render <MessageContainer /> and compare with snapshot 1`] = `
+                  "scaleY": -1,
+                },
+                Object {
+                  "perspective": 1280,
```
Why this?

brunolemos commented:
I found very weird that the code `transform: [{ scaleY: -1 }, { perspective: 1280 }]` appears 6 times in this PR and I'm not sure why they are necessary. Maybe this can be optimized? Implemented differently?

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
Thanks, @brunolemos I removed this style `transform: [{ scaleY: -1 }, { perspective: 1280 }]`

xcarpentier commented:
Thanks @brunocascio, I removed state and extends PureComponent

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

brunolemos commented:
This last one is working fine.

When there are dozens of messages and I scroll to the top fast, it takes some seconds to load the previous items. I was expecting it to be instant as I don't think it's using any server on this example. Or is it?

xcarpentier commented:
No server.
But I need to figured out what is the best configuration for this FlatList.

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
@brunolemos I changed props on FlatList, please try and let me know
```javascript
<FlatList
  initialNumToRender={10}
  maxToRenderPerBatch={10}
  windowSize={10}
/>
```

brunolemos commented:
@xcarpentier Check this:

![gif](https://user-images.githubusercontent.com/619186/35397719-5455c5a0-01d7-11e8-948c-2045c3b3a9ce.gif)

Maybe it's not related to this PR?

xcarpentier commented:
@brunolemos I think it is, thanks you.
Will dig more into it...

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

brunolemos commented on src/MessageContainer.js:
```diff
@@ -107,7 +107,7 @@ export default class MessageContainer extends React.PureComponent {
           maxToRenderPerBatch={MAX_RENDER_ITEM}
           windowSize={MAX_RENDER_ITEM}
-          removeClippedSubviews
+          removeClippedSubviews={false}
```
Did you face any problem? Using false has performance implications.

xcarpentier replied:
The doc mention bug related to this props so I try to see that change but no https://facebook.github.io/react-native/docs/flatlist.html#removeclippedsubviews

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

Shelf-A commented:
Awesome work @xcarpentier.

Thanks for modifying for flatlist. The last PR looks great but everytime the list gets larger the keyboard freezes and the chat input gets slower too and then the chats disappear momentarily when you scroll up . I don't know if this is caused by Expo...

xcarpentier commented:
@Shelf-A thank you very much for your feedbacks.
iOS or Android?

Shelf-A commented:
Android

brunocascio commented:
@xcarpentier do you need some help here?

xcarpentier commented:
@brunocascio YES, I don't find time to figured out what's going on with "android" issues when large list!
We need to find best FlatList props definition I guess?

brunocascio commented:
@xcarpentier OK. I will try to figure out what's going on later 👍

Shelf-A commented:
@xcarpentier @brunocascio I can pinpoint few issues.

1. Constant.js sets a static MAX_RENDER_ITEM to 25 which is used for initialNumToRender and according to https://facebook.github.io/react-native/docs/flatlist.html#initialnumtorender we don't need to render 25 items. maxToRenderPerBatch is MAX_RENDER_ITEM * 2 which gives us 50 which is good for fill rate but performance not so good cos the send button freezes, same goes for windowSize

2. InputToolbar.js this component re-renders many times. It re-renders 7 times when the keyboard is shown and 7 more times when the keyboard is dismissed.  This often affects the root screen when you create chats and navigate back without dismissing the keyboard.

I hope this was helpful.

filbertteo commented on src/MessageContainer.js:
```diff
@@ -95,34 +51,36 @@ export default class MessageContainer extends React.Component {
+      previousMessage,
+      nextMessage,
+      position: item.user._id === this.props.user._id ? 'right' : 'left',
+      hash: sum(
```
What is the purpose of `hash`? As far as I'm aware, the `Message` component does not expect a `hash` prop.

PeterKottas commented:
Hi gents, wow there's a lot of PRs flying around :) I was working on #768 and realized react-native-invertible-scroll-view doesn't work with web. Then I started looking around at ways of fixing it and found that apparently you removed the dependency from this branch altogether. Kind of good news for me I think :) But I wanted to check if my assumptions are correct and also the status of this. How confident do you feel this will end up in master? And I know I am a bit pushy (sorry) but do you have a clue about timelines for this? Thanks!

PeterKottas commented:
Looking at the code here https://github.com/FaridSafi/react-native-gifted-chat/pull/705/commits/10b8e6a9ee183761277f72f85ba2e54560717677#diff-bd8395a0a94eca8b61a34ec66f8a605cR112 I have a bad feeling it might not work anyways. To me, the obvious solution would be to reverse messages order and put a flex-filler on top of the list. However, I am afraid that native implementation might not work with that. Any thoughts?

PeterKottas commented:
That actually happens on current master as well. At least on android.

PeterKottas commented:
Tried again and it actually works ok. The app probably wasn't rebuild properly. It does that on the expo example app though. Not sure what release that points to though.

xcarpentier commented:
@PeterKottas yes, it's difficult to know which version is executed on expo, I think it will be great to add version / branch visible somewhere (header?)

BTW @Shelf-A @PeterKottas Do you think it's a bocker?

PeterKottas commented:
Maybe we could do a commit hash as part of CI and make a link to it on github? We had something like that on a solution long time ago and it proved very popular with testers <-> developers relation.

PeterKottas commented:
Also I am all for merging this asap cause this is a BIG piece of work and I am afraid of merge-hell if we wait too long.

theplatapi commented:
Can this please be merged? I have a chat app out but it slows down after time due to the old list memory issues. Just using the flatlist will be a huge improvement.

brunocascio commented:
@xcarpentier 
I have access to npm. Let me know when it is ready, and I'll deploy it!

xcarpentier commented:
@brunocascio I have access too, but I think it will be great to make a v0.4.0 first, no?
With current master.

brunocascio commented:
@xcarpentier How does it looks? :) https://github.com/FaridSafi/react-native-gifted-chat/releases/tag/v0.4.0

xcarpentier commented:
@brunocascio OK for me

brunocascio commented:
@xcarpentier I've just published v0.4.1 version. Go ahead :)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
@Shelf-A Thanks for your feedback, I changed something:
1. Removed props on Flatlist
1. Added a condition on setState in InputToolbar.js

Can you have a look?

xcarpentier commented:
@PeterKottas I just added the version in header of expo-example. ;)

xcarpentier commented:
@theplatapi I want too but please test it here and let me know if you find any issues: https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

Shelf-A commented:
It starts off perfectly then when the list grows/after sending a couple of messages the input/keyboard freezes as you type and also when you clear text.
![screenshot_20180316-151624](https://user-images.githubusercontent.com/28649534/37522872-fe450ff8-292d-11e8-9f2e-fa0ba39c03cb.png)
Platform: android

xcarpentier commented:
@Shelf-A So, we need to stop rendering of bunch list using FlatList props optimization somehow.

cc https://github.com/facebook/react-native/issues/13413

xcarpentier commented:
Perf optimization in last version:
- removeClippedSubviews on android only
- Message extends PureComponent

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

xcarpentier commented:
@Shelf-A can you test the last one please? 🙄

xcarpentier commented:
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  ![QR Code](https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=https://exp.host/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist)

Shelf-A commented:
@xcarpentier the issue is fixed, works smooth

xcarpentier commented:
Question:

Is that better to remove PureComponent to use shouldComponentUpdate instead because with a small scope conditional update by component (ie. regarding the component and child needs...)

Because PureComponent make a big comparison and it's not worthing for all. Should we consider using more shouldComponentUpdate or keep as is?

Shelf-A commented:
I think we should consider using more shouldComponentUpdate, like for double ticks and time

brunocascio commented:
React devs recommend `PureComponent` only on stateless components. Keep it in mind we can choose between `PureComponent and `Component` depending on how to perform it in your app.
`PureComponent` is just a `Component` that implements `shouldComponentUpdate` with shallow comparison.

xcarpentier commented:
> `PureComponent` is just a `Component` that implements `shouldComponentUpdate` with shallow comparison.

Yes I know! But this is currently my question: why make a big shallow comparison (ie. with PureComponent)? Because in GC we passed all the props from the root to all leaf.

So I think it will be better for us to use `shouldComponentUpdate` to compare only props we need by components and children of this component or change our mind totally by not passing all props to all components.

- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/GiftedChat.js#L314
- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/GiftedChat.js#L436
- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/MessageContainer.js#L120
- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/InputToolbar.js#L68
- etc

So the *shallow comparison* is pretty huge, this is not just from a few props but all props.

What do you think?

(I think you should keep this way to not change to much but instead use more `shouldComponentUpdate` than `PureComponent` for perf reason)

brunocascio commented:
Yes, I agree. I think that performance optimization could take place when those troubles appear. I tested it and it works, so we can release a new version in order to get some user feedbacks as well :)

theplatapi commented:
Tested this branch out with my current app. Rendering a user's message after it's sent takes a little longer, but the app no longer slows down as the list grows. I'll use what's here for now since users are really complaining about the slowdowns!

Also, the load previous button doesn't show the animation when pressed, probably because it was moved to a pure component. Other than that a great improvement!

brunocascio commented:
@xcarpentier The `example`  folder is already updated or what should I use for testing?

helielson commented:
@xcarpentier any updates or ETA here? I'm still using the `modified-for-flatlist` branch 😆

brunolemos commented on src/MessageContainer.js:
```diff
@@ -131,36 +84,32 @@ export default class MessageContainer extends React.Component {
+          contentContainerStyle={styles.contentContainerStyle}
+          renderItem={this.renderRow}
+          renderHeader={this.renderFooter}
+          renderFooter={this.renderLoadEarlier}
```
Why is `renderHeader=renderFooter` and `renderFooter=renderLoadEarlier`? 🤪

howg0924 commented:
@brunolemos as for the TextInput does not move up problem, I fixed it in my own fork, maybe you could take it as a reference: https://github.com/howg0924/react-native-gifted-chat/commit/06e78bee77624d2a8755a3b072bed56208667fb9

I solved it by calling keyboard listeners directly in GiftedChat.js. I don't understand why these listeners were passed down to the FlatList in MessageContainer.js.

The problem only happens on iOS on my app.

brunolemos commented:
@howg0924 unfortunately cherry-picking this commit did not solve the issue... maybe because your branch is outdated?

howg0924 commented:
@brunolemos: My fork is not outdated on this branch. I also fixed some other issues and did a little enhancement on my fork. I guess maybe these modification also helped to fixed the issue. Or maybe the root cause of your problem is different from mine.

brunolemos commented:
@howg0924 yeah I merged master on it, maybe it was that

b4stien commented:
Hello everyone, I am not sure what's blocking this PR, can we help in any way?

We're using this project at my company and would be willing to help land this (which seems to be fixing a rendering issue with messages not appearing until scroll).

xcarpentier commented:
@b4stien some performance issue. Need to add some `suc` and maybe change some pure component

sibelius commented:
@FaridSafi can we merge this, and track some regression in new issues?

xcarpentier commented:
@sibelius I will merge it tomorrow.

helielson commented:
Nice @xcarpentier! Can't wait! Let me know if you need any help

jaburcodes commented:
awesome PR!! Thanks @xcarpentier pretty sure this will help a lot of devs!

dead23angel commented:
bump version pls

varungupta85 commented on src/InputToolbar.js:
```diff
@@ -33,15 +33,19 @@ export default class InputToolbar extends React.Component {
-    this.setState({
-      position: 'relative',
-    });
+    if (this.state !== 'relative') {
```
Should this be `this.state.position`? Same for the next if condition also.

brunolemos replied:
This is probably what caused this bug: https://github.com/FaridSafi/react-native-gifted-chat/pull/705#pullrequestreview-120090347

fbartho commented:
@varungupta85 that bug looks like it just got merged into master!

ChrisEdson commented:
Is it possible to get a release with the new flat list design?

glebez commented:
Is it just me, or `rednerFooter` prop of gifted chat stopped properly working after this branch has been merged?

On first sight it seems to me, that `FlatList` does not expect props `renderHeader` and `renderFooter`, but instead `ListHeaderComponent` and `ListFooterComponent`, which makes them be ignored.

xcarpentier commented:
@glebez Do you have any screenshot or code to share?
Or maybe think about create a PR with the fix...

sibelius commented:
We should have an issue with all possible regressions of this PR

glebez commented:
@xcarpentier, I'll try to put together a reduced test case first and afterwards a PR.

glebez commented:
Here's a test case using current npm package, so before the FlatList was introduced and the footer is displayed properly: https://snack.expo.io/@glebkost/listview-example

If you change the package to use current master the footer is not displayed. (Unfortunately it's not possible in expo snack to set the package to github url, so you need to copy the snack to your local env to test that).
